### PR TITLE
[DOCS] add examples for yield

### PR DESCRIPTION
**Pull request by nicolechung**
## What changed
Added in more examples for `yield`

https://api.emberjs.com/ember/4.4/classes/Ember.Templates.helpers/methods/yield?anchor=yield

- [x] add yield with hash helper
- [x] add yield with hash and component helper
- [x] add yield for multiple block parameters
- [x] add yield with hash and component helper with attributes and args

## Rationale
Coming from React, yield is a bit of a weird concept. Having more examples of how to use `yield` with `hash` and `component` helpers will help people making components who are coming from another framework and are trying to make more composable components with a small API footprint (more "HTML" like components)

Partly inspired by: https://www.youtube.com/watch?v=uTFxv0kReYk&list=PLRtznLaEWGuf87iG3LZarzyMb24JCX9Pv

- wanted to show that you could use @arguments and attributes
- wanted to show that you could move components around

### CODE REVIEW

File: packages/@ember/-internals/glimmer/index.ts [hunk 1]
```diff
@@ -142,6 +142,145 @@
...
+    <banner.Title>Banner title</banner.Title>
+    <banner.Subtitle>Banner subtitle</banner.Subtitle>
+    <banner.Body>A load of body text</banner.Body>
+  </Banner>
+  ```
+
+  ```app/components/Banner.hbs
+  <div>
+    {{yield (hash
+       Title=(component 'path/to/title')
```
Comment (NullVoxPopuli) [inline packages/@ember/-internals/glimmer/index.ts:211]: these paths can be a little wonkey -- do we instead want to promote using component values?

```js
import Title from ''./title";
import Subtitle from ''./subtitle";
import Body from ''./body";

export default class Banner extends Component {
  Title = Title;
  Subtitle = Subtitle;
  Body = Body;
}
```
```hbs
<div>
  {{yield (hash
    Title=this.Title
    Subtitle=this.Subtitle
    Body=(component this.Body defaultArg="some value")
  )}}
</div>
```

**Review by jenweber — APPROVED** [review]
Thank you! It is awesome to have some more examples of what you can do with yield. This fills a gap in our docs overall.

To review this, I looked over the diff and consider whether this belonged here or in the Guides. I believe it belongs here because more involved examples (than usual) are required in order to understand how to use this feature. I appreciate the briefness of the examples as well, given that these are the API docs and not a tutorial.

Next steps are that someone from Framework core should review and merge. Maybe @locks ?

File: packages/@ember/-internals/glimmer/index.ts [hunk 2]
```diff
@@ -142,6 +142,157 @@
...
+  <div>
+    {{yield (hash start=@value.start end=@value.end)}}
+  </div>
+  ```
+
+  Result:
+
+  ```html
+  <div>
+    Start Date: July 1st
```
Comment (bertdeblock) [inline packages/@ember/-internals/glimmer/index.ts:165]:
```suggestion
    Start date: July 1st
```

Nitpick, but the initial template uses a lowercase `d`. Same for `End date`.

File: packages/@ember/-internals/glimmer/index.ts [hunk 3]
```diff
@@ -142,6 +142,157 @@
...
+
+  ```app/templates/application.hbs
+  <Banner @value={{@model}} as |title subtitle body|>
+    <h1>{{title}}</h1>
+    <h2>{{subtitle}}</h2>
+    {{body}}
+  </Banner>
+  ```
+
+  ```app/components/Banner.hbs
```
Comment (bertdeblock) [inline packages/@ember/-internals/glimmer/index.ts:180]:
```suggestion
  ```app/components/banner.hbs
```

File: packages/@ember/-internals/glimmer/index.ts [hunk 4]
```diff
@@ -142,6 +142,157 @@
...
+
+  ```app/templates/application.hbs
+  <Banner @value={{@model}} as |banner|>
+    <banner.Title>Banner title</banner.Title>
+    <banner.Subtitle>Banner subtitle</banner.Subtitle>
+    <banner.Body>A load of body text</banner.Body>
+  </Banner>
+  ```
+
+  ```app/components/banner.ts
```
Comment (bertdeblock) [inline packages/@ember/-internals/glimmer/index.ts:208]:
```suggestion
  ```app/components/banner.js
```

AFAICT, the guides have only been using `.js`?

File: packages/@ember/-internals/glimmer/index.ts [hunk 5]
```diff
@@ -142,6 +142,157 @@
...
+  <Banner @value={{@model}} as |banner|>
+    <banner.Subtitle class="mb-1">Banner subtitle</banner.Subtitle>
+    <banner.Title @variant="loud">Banner title</banner.Title>
+    <banner.Body>A load of body text</banner.Body>
+  </Banner>
+  ```
+
+  ```app/components/banner/subtitle.hbs
+  {{!-- note the use of ..attributes --}}
+  <div ...attributes>
```
Comment (bertdeblock) [inline packages/@ember/-internals/glimmer/index.ts:273]:
```suggestion
  <h2 ...attributes>
```

As this component has been rendering an `h2` tag in the previous examples.

File: packages/@ember/-internals/glimmer/index.ts [hunk 6]
```diff
@@ -142,6 +142,157 @@
...
+
+  ```app/templates/application.hbs
+  <Banner @value={{@model}} as |banner|>
+    <banner.Subtitle class="mb-1">Banner subtitle</banner.Subtitle>
+    <banner.Title @variant="loud">Banner title</banner.Title>
+    <banner.Body>A load of body text</banner.Body>
+  </Banner>
+  ```
+
+  ```app/components/banner/subtitle.hbs
```
Comment (bertdeblock) [inline packages/@ember/-internals/glimmer/index.ts:271]: The file path doesn't match the import statement in the previous example.

The banner component is located at `app/components/banner.js`.

The import statement for the subtitle component is `import Subtitle from './subtitle';`, which means the subtitle component is actually located at `app/components/subtitle.hbs`.

We could also update the import statements to:
```js
import Title from './banner/title';
import Subtitle from './banner/subtitle';
import Body from './banner/body';
```

File: packages/@ember/-internals/glimmer/index.ts [hunk 7]
```diff
@@ -142,6 +142,157 @@
...
+  {{!-- note the use of ..attributes --}}
+  <div ...attributes>
+    {{yield}}
+  </div>
+  ```
+
+  ```app/components/banner/title.hbs
```
Comment (bertdeblock) [inline packages/@ember/-internals/glimmer/index.ts:278]: Same remark about the file path.

File: packages/@ember/-internals/glimmer/index.ts [hunk 8]
```diff
@@ -142,6 +142,157 @@
...
+  {{#if @variant "loud"}}
```
Comment (bertdeblock) [inline packages/@ember/-internals/glimmer/index.ts:279]:
```suggestion
  {{#if (eq @variant "loud")}}
```

The current example throws a syntax error.

File: packages/@ember/-internals/glimmer/index.ts (hunk 6 above)
Comment (nicolechung) [reply packages/@ember/-internals/glimmer/index.ts:271]: I changed it to the latter to indicate that these components are to be used inside of "banner"

File: packages/@ember/-internals/glimmer/index.ts (hunk 8 above)
Comment (nicolechung) [reply packages/@ember/-internals/glimmer/index.ts]: Thanks for catching this!

Comment (jelhan) [reply packages/@ember/-internals/glimmer/index.ts]: Truth helpers haven't been shipped yet, have they? Using a helper, which is not available by default, might be confusing.

Comment (NullVoxPopuli) [reply packages/@ember/-internals/glimmer/index.ts]: Let's get them shipped 🙃 

Cc @chancancode 🥳

**Comment by nicolechung** [comment]
Question: I can't seem to `re-request` review or resolve comments? - but if anyone wants to (re) take a look, that would be appreciated!

**Comment by nicolechung** [comment]
Question: Should (I also) do a comparison with `let` versus `yield` here?

let with hash + component helper: have to render the components in the same file
yield with hash + component helper: allows the consumer of the component render how they want with args and attributes

File: packages/@ember/-internals/glimmer/index.ts (hunk 1 above)
Comment (chriskrycho) [reply packages/@ember/-internals/glimmer/index.ts]: I *don’t* think we should do that. Among other things, that creates a per-instance binding, which isn’t a lot but also isn’t necessary or great for performance; and in this case it introduces a backing class for *only* that. Using the actual paths is better until we switch to teaching `<template>` and can just show yielding the values normally.

**Review by bertdeblock — APPROVED** [review]

**Comment by jenweber** [comment]
Hi Nicole, thanks again for your work on this!

I recommend that the framework team does a final review & merge this, and any additional improvements can be a followup PR.

I discussed the issue of Guides vs API docs as the home for this with locks. After that conversation, I think eventually it would be good to show examples like this in a cookbook or “advanced” section of the guides. If anyone wants to give that a try, in addition to this PR moving forward, that’s welcome! We have room for more commentary there.

**Review by mixonic — APPROVED** [review]

**Comment by mixonic** [comment]
Thank you! This seems reasonable to me.

**Pull request merged**